js/gestionPresupuesto.js

```diff
@@ -1,25 +1,98 @@
-// TODO: Crear las funciones, objetos y variables indicadas en el enunciado
+let presupuesto = 0;
+let gastos = [];
+let idGasto = 0;
 
-// TODO: Variable global
+let comprobarPositivo = (numero) => Number.isFinite(numero) && numero >= 0;
+let ComprobarFecha = (fecha) => Date.parse(fecha);
 
-
-function actualizarPresupuesto() {
-    // TODO
+function actualizarPresupuesto(valor) {
+  if (comprobarPositivo(valor)) return (presupuesto = valor);
+  else {
+    console.log("Valor no válida");
+    return -1;
+  }
 }
 
 function mostrarPresupuesto() {
-    // TODO
+  return `Tu presupuesto actual es de ${presupuesto} €`;
+}
+
+function listarGastos() {
+  return gastos;
+}
+
+function anyadirGasto(gasto) {
+  gasto.id = idGasto;
+  idGasto++;
+  gastos.push(gasto);
 }
 
-function CrearGasto() {
-    // TODO
+function borrarGasto(id) {
+  gastos = gastos.filter((gasto) => gasto.id !== id);
 }
 
-// NO MODIFICAR A PARTIR DE AQUÍ: exportación de funciones y objetos creados para poder ejecutar los tests.
-// Las funciones y objetos deben tener los nombres que se indican en el enunciado
-// Si al obtener el código de una práctica se genera un conflicto, por favor incluye todo el código que aparece aquí debajo
-export   {
-    mostrarPresupuesto,
-    actualizarPresupuesto,
-    CrearGasto
+function calcularTotalGastos() {
+  return gastos.reduce((total, gasto) => (total += gasto.valor), 0);
 }
+
+function calcularBalance() {
+  return presupuesto - calcularTotalGastos();
+}
+
+function CrearGasto(descripcion, valor, fecha, ...etiquetas) {
+  this.descripcion = descripcion;
+  this.valor = comprobarPositivo(valor) ? valor : 0;
+  this.fecha = ComprobarFecha(fecha) || new Date();
+  this.etiquetas = [...etiquetas];
```

Comment:
En lugar de añadirlas directamente, puedes utilizar `this.anyadirEtiquetas`. De esta manera comprobarás que no introduces duplicados.

Reply:
¿Así?
`this.etiquetas = this.anyadirEtiquetas(...newEtiquetas);`

Reply:
Eso es 👏